File: java/sorting/InsertionSortProduction.java
import java.util.Comparator;
import java.util.List;

public final class InsertionSortProduction {
    private InsertionSortProduction() {
    }

    public static <T extends Comparable<? super T>> void binaryInsertionSortInPlace(List<T> values) {
        binaryInsertionSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void binaryInsertionSortInPlace(List<T> values, Comparator<? super T> comparator) {
        if (values == null || comparator == null) {
            throw new IllegalArgumentException("values and comparator must not be null");
        }

        for (int index = 1; index < values.size(); index++) {
            T current = values.get(index);
            int insertAt = upperBound(values, current, 0, index, comparator);

            if (insertAt == index) {
                continue;
            }

            for (int move = index; move > insertAt; move--) {
                values.set(move, values.get(move - 1));
            }

            values.set(insertAt, current);
        }
    }

    private static <T> int upperBound(List<T> values, T target, int low, int high, Comparator<? super T> comparator) {
        while (low < high) {
            int mid = low + (high - low) / 2;

            if (comparator.compare(values.get(mid), target) <= 0) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }
}

File: python/sorting/insertion_sort_production.py
def binary_insertion_sort_in_place(values, *, key=None, reverse=False):
    key_fn = key or (lambda value: value)

    for index in range(1, len(values)):
        current = values[index]
        current_key = key_fn(current)
        insert_at = _upper_bound(values, current_key, 0, index, key_fn, reverse)

        if insert_at == index:
            continue

        values[insert_at + 1:index + 1] = values[insert_at:index]
        values[insert_at] = current

    return values


def _upper_bound(values, target_key, low, high, key_fn, reverse):
    while low < high:
        mid = low + (high - low) // 2
        mid_key = key_fn(values[mid])

        if mid_key >= target_key if reverse else mid_key <= target_key:
            low = mid + 1
        else:
            high = mid

    return low
